File: CSharpCodeBase/components/character/baseweapon.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class baseweapon {
 public void init (self, entity, name);
   this.name = name;
   this.entity = entity;
   this.config = ConfigComponent("weapons", this.name);
   this.minDistance = this.config:Get("minDistance")  ||  0;
   this.maxDistance = this.config:Get("maxDistance");
   this.midDistance = this.config:Get("midDistance")  ||  ((this.maxDistance - this.minDistance) / 2);
   this.attackTimeout = this.config:Get("attackTimeout");
   this.attackPrepareTimeout = this.config:Get("attackPrepareTimeout");
   this.currentComboState = 1;
   this.remote = this.config:Get("remote")  ||  0;
   
   var actionTypes = {}
   var combo = this.config:Get("combo");
   var timing = this.config:Get("combotimeout");
   var effects = this.config:Get("comboeffect")  ||  {}
   
   for(k, v in pairs(combo) ){
     actionTypes[#actionTypes + 1] = {v, timing[k], effects[k]}
   }
   this.actionTypes = actionTypes;
   
   // Load damage parameters
   self:LoadDamage();
 }})
 public void LoadDamage(){
   this.criticalChance = this.config:Get("criticalChance");
   this.damage = {}
   var damageString = this.config:Get("damage");
   for(dType, dValue in string.gmatch(damageString, "(%w+)%:(%d+)%,?") ){
     this.damage[dType] = dValue;
   }
 }
 public void GetCriticalMultiplier(){
   return math.chance(this.criticalChance)  &&  2  ||  1;
 }
 public void GetDamage(){
   var damage = {}
   var critMul = self:GetCriticalMultiplier();
   for(k, v in pairs(this.damage) ){
     damage[k] = v * critMul;
   }
   var effect = this.actionTypes[this.currentComboState][3];
   return {damage=damage, effects={critical=(critMul > 1  &&  true  ||  false), punch=effect}}
 }
 public void CanAttack(target){
   var effectiveDistance = this.entity:GetEffectiveDistance(target);
   // return effectiveDistance > this.minDistance  &&  effectiveDistance < this.maxDistance
   return effectiveDistance < this.maxDistance  &&  (this.minDistance <= 0  ||  effectiveDistance > this.minDistance  ||  this.remote);
 }
 public void GetActionTypes(target){
   return this.actionTypes;
 }
 public void GetAttackDistance(){
   return this.midDistance //(this.maxDistance - this.minDistance) / 2
 }
 public void GetAttackTimeout(){
   return this.attackTimeout  ||  0;
 }
 public void GetAttackPrepareTimeout(){
   return this.attackPrepareTimeout  ||  0;
 }
 public void OnActivate(){
   var transforms = this.config:Get("transform");
   if(transforms != null  ){ 
     for(k, v in pairs(transforms) ){
       Transform.SetRendererState(this.entity.transform, v, true);
     }
   }
 }
 public void OnDeactivate(){
   var transforms = this.config:Get("transform");
   if(transforms != null  ){ 
     for(k, v in pairs(transforms) ){
       Transform.SetRendererState(this.entity.transform, v, false);
     }
   }
 }
 public void Attack(target){
   if(this.remote > 0  ){ 
     // throw bullet
     var bullet = BulletEntity("bullet", this.entity, target.GetPosition  &&  target:GetPosition()  ||  target, self);
     GameController:AddEntity(bullet);
     //print("Throw bullet!")
     // create bullet weapon
   }else{
     this.entity.damageProcessor:SendDamage(target, self:GetDamage())  ;
   }
 }
 }}

File: CSharpCodeBase/components/character/baseweapon.lua
local Class = require 'utils.hump.class'
local ConfigComponent = require 'components.services.config'
local BulletEntity = require 'entities.weapon.bullet'

local BaseWeapon = Class({init = function (self, entity, name)
  self.name = name
  self.entity = entity
  self.config = ConfigComponent('weapons', self.name)
  self.minDistance = self.config:Get('minDistance') or 0
  self.maxDistance = self.config:Get('maxDistance')
  self.midDistance = self.config:Get('midDistance') or ((self.maxDistance - self.minDistance) / 2)
  self.attackTimeout = self.config:Get('attackTimeout')
  self.attackPrepareTimeout = self.config:Get('attackPrepareTimeout')
  self.currentComboState = 1
  self.remote = self.config:Get('remote') or 0
  
  local actionTypes = {}
  local combo = self.config:Get('combo')
  local timing = self.config:Get('combotimeout')
  local effects = self.config:Get('comboeffect') or {}
  
  for k, v in pairs(combo) do
    actionTypes[#actionTypes + 1] = {v, timing[k], effects[k]}
  end
  self.actionTypes = actionTypes
  
  -- Load damage parameters
  self:LoadDamage()
end})

function BaseWeapon:LoadDamage()
  self.criticalChance = self.config:Get('criticalChance')
  self.damage = {}
  local damageString = self.config:Get('damage')
  for dType, dValue in string.gmatch(damageString, '(%w+)%:(%d+)%,?') do
    self.damage[dType] = dValue
  end
end

function BaseWeapon:GetCriticalMultiplier()
  return math.chance(self.criticalChance) and 2 or 1
end

function BaseWeapon:GetDamage()
  local damage = {}
  local critMul = self:GetCriticalMultiplier()
  for k, v in pairs(self.damage) do
    damage[k] = v * critMul
  end
  local effect = self.actionTypes[self.currentComboState][3]
  return {damage=damage, effects={critical=(critMul > 1 and true or false), punch=effect}}
end

function BaseWeapon:CanAttack(target)
  local effectiveDistance = self.entity:GetEffectiveDistance(target)
  -- return effectiveDistance > self.minDistance and effectiveDistance < self.maxDistance
  return effectiveDistance < self.maxDistance and (self.minDistance <= 0 or effectiveDistance > self.minDistance or self.remote)
end

function BaseWeapon:GetActionTypes(target)
  return self.actionTypes
end

function BaseWeapon:GetAttackDistance()
  return self.midDistance --(self.maxDistance - self.minDistance) / 2
end

function BaseWeapon:GetAttackTimeout()
  return self.attackTimeout or 0
end

function BaseWeapon:GetAttackPrepareTimeout()
  return self.attackPrepareTimeout or 0
end

function BaseWeapon:OnActivate()
  local transforms = self.config:Get('transform')
  if transforms ~= nil then
    for k, v in pairs(transforms) do
      Transform.SetRendererState(self.entity.transform, v, true)
    end
  end
end

function BaseWeapon:OnDeactivate()
  local transforms = self.config:Get('transform')
  if transforms ~= nil then
    for k, v in pairs(transforms) do
      Transform.SetRendererState(self.entity.transform, v, false)
    end
  end
end

function BaseWeapon:Attack(target)
  if self.remote > 0 then
    -- throw bullet
    local bullet = BulletEntity('bullet', self.entity, target.GetPosition and target:GetPosition() or target, self)
    GameController:AddEntity(bullet)
    --print("Throw bullet!")
    -- create bullet weapon
  else
    self.entity.damageProcessor:SendDamage(target, self:GetDamage())  
  end
end

return BaseWeapon